src/lsm/groove.zig

```diff
@@ -640,67 +719,120 @@ pub fn GrooveType(
 
             groove.prefetch_snapshot = snapshot_target;
             assert(groove.prefetch_keys.count() == 0);
+
+            if (has_id) groove.timestamps.reset();
         }
 
         /// This must be called by the state machine for every key to be prefetched.
         /// We tolerate duplicate IDs enqueued by the state machine.
         /// For example, if all unique operations require the same two dependencies.
         pub fn prefetch_enqueue(groove: *Groove, key: PrimaryKey) void {
+            // No need to check again if the key is already present.
+            if (groove.prefetch_keys.contains(.{ .id = key })) return;
+
             if (has_id) {
                 if (!groove.ids.key_range_contains(groove.prefetch_snapshot.?, key)) return;
 
                 if (groove.objects_cache.has(key)) {
                     return;
                 }
 
-                groove.prefetch_from_memory_by_id(key);
+                groove.prefetch_from_memory_by_id(key, .objects_cache);
             } else {
                 if (groove.objects_cache.has(key)) {
                     return;
                 }
 
-                groove.prefetch_from_memory_by_timestamp(key);
+                groove.prefetch_from_memory_by_timestamp(key, .objects_cache);
             }
         }
 
+        /// This must be called by the state machine for every timestamp to be checked by `exists`.
+        /// The first call to this function may trigger the sorting of the mutable table, which is
+        /// likely a no-op since timestamps are strictly increasing and the table should already
+        /// be sorted, except for objects that are frequently updated (e.g., accounts).
+        /// We tolerate duplicate timestamps enqueued by the state machine.
+        pub fn prefetch_exists_enqueue(
+            groove: *Groove,
+            timestamp: u64,
+        ) void {
+            // Only applicable to objects with an `id` field.
+            // Use `prefetch_enqueue` if the object is already keyed by timestamp.
+            comptime assert(has_id);
+
+            // No need to check again if the timestamp is already present.
+            if (!groove.timestamps.enroll(timestamp)) return;
+
+            // The mutable table needs to be sorted to enable searching by timestamp.
+            // The immutable table will be searched by `prefetch_from_memory_by_timestamp`.
+            groove.objects.table_mutable.sort();
+            if (groove.objects.table_mutable.get(timestamp)) |object| {
+                assert(object.timestamp == timestamp);
+                groove.timestamps.found(timestamp);
+                return;
+            }
+
+            groove.prefetch_from_memory_by_timestamp(timestamp, .timestamps);
```

Comment:
> Instead of a Set, the TimestampSet it's backed by a HashMap(u64,bool) precisely to assert that we can only call exists for timestamps that were enrolled before.

Right now we enroll/enqueue the timestamp immediately, before we know whether or not it exists. I was suggesting that we only insert the timestamp (along with `true` or `false`) once we know for sure.

With the existing code, if the prefetch result ends up in the map instead of the set, `.exists()` will return `false` instead of panicking, since we enrolled it immediately in `prefetch_exists_enqueue`.

Reply:
Done! We should probably do something similar with `prefetch_enqueue`, asserting that the key passed in `groove.get` was previously enqueued for prefetching.
I'll send it in a follow-up PR.